Mathlib/RingTheory/Binomial.lean

```diff
@@ -176,6 +176,98 @@ instance Int.instBinomialRing : BinomialRing ℤ where
+      show (-1 + -↑n = (-1 : ℤ) * (n + 1)) by omega, ← mul_assoc, pow_add, pow_one,
+      Nat.factorial, Nat.cast_mul, ← mul_assoc, Nat.cast_succ]
+
+@[simp]
+theorem ascPochhammer_succ_smeval_neg (n : ℕ) :
+    smeval (ascPochhammer ℕ (n + 1)) (-n : ℤ) = 0 := by
+  rw [ascPochhammer_succ_right, smeval_mul, smeval_add, smeval_X, ← C_eq_natCast, smeval_C,
+    pow_zero, pow_one, Nat.cast_id, nsmul_eq_mul, mul_one, add_left_neg, mul_zero]
+
+theorem ascPochhammer_smeval_neg_add (n : ℕ) : ∀(k : ℕ),
+    smeval (ascPochhammer ℕ (n + k + 1)) (-n : ℤ) = 0
+  | 0 => by
+    rw [add_zero, ascPochhammer_succ_smeval_neg]
+  | k + 1 => by
+    rw [ascPochhammer_succ_right, smeval_mul, ← add_assoc, ascPochhammer_smeval_neg_add n k,
+      zero_mul]
+
+@[simp]
+theorem ascPochhammer_smeval_neg_lt (n k : ℕ) (h : n < k) :
+    smeval (ascPochhammer ℕ k) (-n : ℤ) = 0 := by
+  have hk : k = n + (k - n - 1) + 1 := by
+    rw [add_rotate, Nat.sub_sub, Nat.add_right_comm, Nat.add_assoc, Nat.sub_add_cancel h]
+  rw [hk, ascPochhammer_smeval_neg_add]
+
+theorem ascPochhammer_smeval_nat_cast [NatPowAssoc R] (n k : ℕ) :
+    smeval (ascPochhammer ℕ k) (n : R) = smeval (ascPochhammer ℕ k) n := by
+  rw [smeval_at_natCast (ascPochhammer ℕ k) n]
+
+theorem multichoose_neg (n : ℕ) : multichoose (-n : ℤ) n = (-1)^n := by
+    refine @nsmul_right_injective ℤ _ _ _ (Nat.factorial n) (Nat.factorial_ne_zero n)
+      (multichoose (-n : ℤ) n) ((-1)^n) ?_
+    simp only
+    rw [factorial_nsmul_multichoose_eq_ascPochhammer, ascPochhammer_smeval_neg_self, nsmul_eq_mul,
+      Nat.cast_comm]
+
+@[simp]
+theorem multichoose_succ_neg (n : ℕ) : multichoose (-n : ℤ) (n + 1) = 0 := by
+  refine @nsmul_right_injective ℤ _ _ _ (Nat.factorial (n + 1)) (Nat.factorial_ne_zero (n + 1))
+    (multichoose (-n : ℤ) (n + 1)) 0 ?_
+  simp only
+  rw [factorial_nsmul_multichoose_eq_ascPochhammer, ascPochhammer_succ_smeval_neg, smul_zero]
+
+theorem multichoose_neg_add (n k : ℕ) : multichoose (-n : ℤ) (n + k + 1) = 0 := by
+  refine nsmul_right_injective (Nat.factorial (n + k + 1)) (Nat.factorial_ne_zero (n + k + 1)) ?_
+  simp only
+  rw [factorial_nsmul_multichoose_eq_ascPochhammer, ascPochhammer_smeval_neg_add, smul_zero]
+
+@[simp]
+theorem multichoose_neg_lt (n k : ℕ) (h : n < k) : multichoose (-n : ℤ) k = 0 := by
+  refine nsmul_right_injective (Nat.factorial k) (Nat.factorial_ne_zero k) ?_
+  simp only
+  rw [factorial_nsmul_multichoose_eq_ascPochhammer, ascPochhammer_smeval_neg_lt n k h, smul_zero]
+
+theorem multichoose_succ_neg_natCast [NatPowAssoc R] (n : ℕ) :
+    multichoose (-n : R) (n + 1) = 0 := by
+  refine nsmul_right_injective (Nat.factorial (n + 1)) (Nat.factorial_ne_zero (n + 1)) ?_
+  simp only [smul_zero]
+  rw [factorial_nsmul_multichoose_eq_ascPochhammer, smeval_at_neg_nat,
+    ascPochhammer_succ_smeval_neg n, Int.cast_zero]
+
+theorem ascPochhammer_smeval_nat_int [NatPowAssoc R] (r : R) : ∀(n : ℕ),
```

Comment:
Perhaps should be `smeval_ascPochhammer_int_ofNat` or `smeval_ascPochhammer_int_coe_nat`.